🐛 Bug Report: `createPermissionRule:toQuery` requires casting (1.36.0->)

### 📜 Description

`toQuery` in `createPermissionRule` requires casting to filter property to prevent TSC error.

### 👍 Expected behavior

It should work without specific casting:

```ts
  toQuery: ({ type }) => {
    return {
      property: 'posts.type' as PostFilter['property'],
      values: [type],
    };
  },
```

### 👎 Actual Behavior with Screenshots

Fails tsc:

```sh
plugins/qeta-node/src/permissionRules.ts:99:3 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(rule: CreatePermissionRuleOptions<Post, PostFilter, PostFilter, "post", { type: string; }>): PermissionRule<Post, PostFilter, "post", { ...; }>', gave the following error.
    Type '({ type }: { type: string; }) => { property: string; values: string[]; }' is not assignable to type '(params: { type: string; }) => PermissionCriteria<PostFilter>'.
      Type '{ property: string; values: string[]; }' is not assignable to type 'PermissionCriteria<PostFilter>'.
        Type '{ property: string; values: string[]; }' is not assignable to type 'PostFilter'.
          Types of property 'property' are incompatible.
            Type 'string' is not assignable to type '"posts.author" | "posts.id" | "tags" | "entityRefs" | "posts.type"'.

99   toQuery: ({ type }) => {
     ~~~~~~~
```

### 👟 Reproduction steps

1. Create permission rule:

```ts
export const postHasType = createPermissionRule({
  name: 'HAS_TYPE',
  description: 'Should allow only if the post has the specific type',
  resourceRef: postPermissionResourceRef,
  paramsSchema: z.object({
    type: z.string().describe('Type to match the post'),
  }),
  apply: (resource: Post, { type }) => {
    return resource?.type === type;
  },
  toQuery: ({ type }) => {
    return {
      property: 'posts.type',
      values: [type],
    };
  },
});
```
2. Run `yarn tsc`

### 📃 Provide the context for the Bug.

_No response_

### 🖥️ Your Environment

OS:   Darwin 24.3.0 - darwin/arm64
node: v22.9.0
yarn: 4.5.0
cli:  0.30.0 (installed)
backstage:  1.36.1

Dependencies:
  @backstage/app-defaults                                          1.5.17
  @backstage/backend-app-api                                       1.2.0
  @backstage/backend-common                                        0.25.0
  @backstage/backend-defaults                                      0.8.1
  @backstage/backend-dev-utils                                     0.1.5
  @backstage/backend-openapi-utils                                 0.5.0
  @backstage/backend-plugin-api                                    1.0.0, 1.2.0
  @backstage/backend-test-utils                                    1.3.0
  @backstage/catalog-client                                        1.7.0, 1.9.1
  @backstage/catalog-model                                         1.7.0, 1.7.3
  @backstage/cli-common                                            0.1.14, 0.1.15
  @backstage/cli-node                                              0.2.13
  @backstage/cli                                                   0.30.0
  @backstage/config-loader                                         1.9.1, 1.9.6
  @backstage/config                                                1.2.0, 1.3.2
  @backstage/core-app-api                                          1.15.5
  @backstage/core-compat-api                                       0.3.6
  @backstage/core-components                                       0.16.4
  @backstage/core-plugin-api                                       1.10.4
  @backstage/dev-utils                                             1.1.7
  @backstage/errors                                                1.2.4, 1.2.7
  @backstage/eslint-plugin                                         0.1.10
  @backstage/frontend-app-api                                      0.10.5
  @backstage/frontend-defaults                                     0.1.6
  @backstage/frontend-plugin-api                                   0.9.5
  @backstage/frontend-test-utils                                   0.2.6
  @backstage/integration-aws-node                                  0.1.12, 0.1.15
  @backstage/integration-react                                     1.2.4
  @backstage/integration                                           1.15.0, 1.16.1
  @backstage/plugin-app                                            0.1.6
  @backstage/plugin-auth-backend-module-atlassian-provider         0.4.0
  @backstage/plugin-auth-backend-module-auth0-provider             0.2.0
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.4.0
  @backstage/plugin-auth-backend-module-azure-easyauth-provider    0.2.5
  @backstage/plugin-auth-backend-module-bitbucket-provider         0.3.0
  @backstage/plugin-auth-backend-module-bitbucket-server-provider  0.2.0
  @backstage/plugin-auth-backend-module-cloudflare-access-provider 0.4.0
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.4.0
  @backstage/plugin-auth-backend-module-github-provider            0.3.0
  @backstage/plugin-auth-backend-module-gitlab-provider            0.3.0
  @backstage/plugin-auth-backend-module-google-provider            0.3.0
  @backstage/plugin-auth-backend-module-guest-provider             0.2.5
  @backstage/plugin-auth-backend-module-microsoft-provider         0.3.0
  @backstage/plugin-auth-backend-module-oauth2-provider            0.4.0
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.2.5
  @backstage/plugin-auth-backend-module-oidc-provider              0.4.0
  @backstage/plugin-auth-backend-module-okta-provider              0.2.0
  @backstage/plugin-auth-backend-module-onelogin-provider          0.3.0
  @backstage/plugin-auth-backend                                   0.24.3
  @backstage/plugin-auth-node                                      0.5.2, 0.6.0
  @backstage/plugin-catalog-backend                                1.31.0
  @backstage/plugin-catalog-common                                 1.1.3
  @backstage/plugin-catalog-node                                   1.16.0
  @backstage/plugin-catalog-react                                  1.15.2
  @backstage/plugin-catalog                                        1.27.0
  @backstage/plugin-events-backend                                 0.4.3
  @backstage/plugin-events-node                                    0.4.8
  @backstage/plugin-home-react                                     0.1.23
  @backstage/plugin-home                                           0.8.5
  @backstage/plugin-notifications-backend                          0.5.3
  @backstage/plugin-notifications-common                           0.0.8
  @backstage/plugin-notifications-node                             0.2.12
  @backstage/plugin-notifications                                  0.5.2
  @backstage/plugin-permission-backend                             0.5.54
  @backstage/plugin-permission-common                              0.8.1, 0.8.4
  @backstage/plugin-permission-node                                0.8.8
  @backstage/plugin-permission-react                               0.4.31
  @backstage/plugin-scaffolder-common                              1.5.9
  @backstage/plugin-scaffolder-node                                0.7.0
  @backstage/plugin-search-backend-module-catalog                  0.3.1
  @backstage/plugin-search-backend-module-pg                       0.5.41
  @backstage/plugin-search-backend-node                            1.3.8
  @backstage/plugin-search-backend                                 1.8.2
  @backstage/plugin-search-common                                  1.2.17
  @backstage/plugin-search-react                                   1.8.6
  @backstage/plugin-search                                         1.4.23
  @backstage/plugin-signals-backend                                0.3.1
  @backstage/plugin-signals-node                                   0.1.17
  @backstage/plugin-signals-react                                  0.0.10
  @backstage/plugin-signals                                        0.0.16
  @backstage/release-manifests                                     0.0.12
  @backstage/test-utils                                            1.7.5
  @backstage/theme                                                 0.6.4
  @backstage/types                                                 1.1.1, 1.2.1
  @backstage/version-bridge                                        1.0.11


### 👀 Have you spent some time to check if this bug has been raised before?

- [x] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [x] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

--- a/.changeset/tidy-ads-cover.md
+++ b/.changeset/tidy-ads-cover.md
@@ -1,0 +1,5 @@
+---
+'@backstage/plugin-permission-node': patch
+---
+
+Improved type inference when passing a `PermissionResourceRef` to `createPermissionRule`.

--- a/plugins/permission-node/report.api.md
+++ b/plugins/permission-node/report.api.md
@@ -201,20 +201,16 @@ export function createPermissionResourceRef<TResource, TQuery>(): {
 
 // @public
 export function createPermissionRule<
-  TResource,
-  TQuery,
-  TQueryOutput extends TQuery,
-  TResourceType extends string,
+  TRef extends PermissionResourceRef,
   TParams extends PermissionRuleParams = undefined,
 >(
-  rule: CreatePermissionRuleOptions<
-    TResource,
-    TQuery,
-    TQueryOutput,
-    TResourceType,
-    TParams
-  >,
-): PermissionRule<TResource, TQuery, TResourceType, TParams>;
+  rule: CreatePermissionRuleOptions<TRef, TParams>,
+): PermissionRule<
+  TRef['TResource'],
+  TRef['TQuery'],
+  TRef['resourceType'],
+  TParams
+>;
 
 // @public @deprecated
 export function createPermissionRule<
@@ -228,19 +224,18 @@ export function createPermissionRule<
 
 // @public (undocumented)
 export type CreatePermissionRuleOptions<
-  TResource,
-  TQuery,
-  TQueryOutput extends TQuery,
-  TResourceType extends string,
+  TRef extends PermissionResourceRef,
   TParams extends PermissionRuleParams,
-> = {
-  name: string;
-  description: string;
-  resourceRef: PermissionResourceRef<TResource, TQuery, TResourceType>;
-  paramsSchema?: z.ZodSchema<TParams>;
-  apply(resource: TResource, params: NoInfer_2<TParams>): boolean;
-  toQuery(params: NoInfer_2<TParams>): PermissionCriteria<TQueryOutput>;
-};
+> = TRef extends PermissionResourceRef<infer IResource, infer IQuery, any>
+  ? {
+      name: string;
+      description: string;
+      resourceRef: TRef;
+      paramsSchema?: z.ZodSchema<TParams>;
+      apply(resource: IResource, params: NoInfer_2<TParams>): boolean;
+      toQuery(params: NoInfer_2<TParams>): PermissionCriteria<IQuery>;
+    }
+  : never;
 
 // @public
 export const isAndCriteria: <T>(

--- a/plugins/permission-node/src/integration/createPermissionRule.ts
+++ b/plugins/permission-node/src/integration/createPermissionRule.ts
@@ -27,36 +27,35 @@ import { NoInfer } from './util';
  * @public
  */
 export type CreatePermissionRuleOptions<
-  TResource,
-  TQuery,
-  TQueryOutput extends TQuery,
-  TResourceType extends string,
+  TRef extends PermissionResourceRef,
   TParams extends PermissionRuleParams,
-> = {
-  name: string;
-  description: string;
+> = TRef extends PermissionResourceRef<infer IResource, infer IQuery, any>
+  ? {
+      name: string;
+      description: string;
 
-  resourceRef: PermissionResourceRef<TResource, TQuery, TResourceType>;
+      resourceRef: TRef;
 
-  /**
-   * A ZodSchema that reflects the structure of the parameters that are passed to
-   */
-  paramsSchema?: z.ZodSchema<TParams>;
+      /**
+       * A ZodSchema that reflects the structure of the parameters that are passed to
+       */
+      paramsSchema?: z.ZodSchema<TParams>;
 
-  /**
-   * Apply this rule to a resource already loaded from a backing data source. The params are
-   * arguments supplied for the rule; for example, a rule could be `isOwner` with entityRefs as the
-   * params.
-   */
-  apply(resource: TResource, params: NoInfer<TParams>): boolean;
+      /**
+       * Apply this rule to a resource already loaded from a backing data source. The params are
+       * arguments supplied for the rule; for example, a rule could be `isOwner` with entityRefs as the
+       * params.
+       */
+      apply(resource: IResource, params: NoInfer<TParams>): boolean;
 
-  /**
-   * Translate this rule to criteria suitable for use in querying a backing data store. The criteria
-   * can be used for loading a collection of resources efficiently with conditional criteria already
-   * applied.
-   */
-  toQuery(params: NoInfer<TParams>): PermissionCriteria<TQueryOutput>;
-};
+      /**
+       * Translate this rule to criteria suitable for use in querying a backing data store. The criteria
+       * can be used for loading a collection of resources efficiently with conditional criteria already
+       * applied.
+       */
+      toQuery(params: NoInfer<TParams>): PermissionCriteria<IQuery>;
+    }
+  : never;
 
 /**
  * Helper function to create a {@link PermissionRule} for a specific resource type using a {@link PermissionResourceRef}.
@@ -64,20 +63,16 @@ export type CreatePermissionRuleOptions<
  * @public
  */
 export function createPermissionRule<
-  TResource,
-  TQuery,
-  TQueryOutput extends TQuery,
-  TResourceType extends string,
+  TRef extends PermissionResourceRef,
   TParams extends PermissionRuleParams = undefined,
 >(
-  rule: CreatePermissionRuleOptions<
-    TResource,
-    TQuery,
-    TQueryOutput,
-    TResourceType,
-    TParams
-  >,
-): PermissionRule<TResource, TQuery, TResourceType, TParams>;
+  rule: CreatePermissionRuleOptions<TRef, TParams>,
+): PermissionRule<
+  TRef['TResource'],
+  TRef['TQuery'],
+  TRef['resourceType'],
+  TParams
+>;
 /**
  * Helper function to ensure that {@link PermissionRule} definitions are typed correctly.
  *
@@ -95,22 +90,19 @@ export function createPermissionRule<
 export function createPermissionRule<
   TResource,
   TQuery,
-  TQueryOutput extends TQuery,
   TResourceType extends string,
+  TRef extends PermissionResourceRef,
   TParams extends PermissionRuleParams = undefined,
 >(
   rule:
     | PermissionRule<TResource, TQuery, TResourceType, TParams>
-    | CreatePermissionRuleOptions<
-        TResource,
-        TQuery,
-        TQueryOutput,
-        TResourceType,
-        TParams
-      >,
+    | CreatePermissionRuleOptions<TRef, TParams>,
 ): PermissionRule<TResource, TQuery, TResourceType, TParams> {
   if ('resourceRef' in rule) {
-    return { ...rule, resourceType: rule.resourceRef.resourceType };
+    return {
+      ...rule,
+      resourceType: rule.resourceRef.resourceType as TResourceType,
+    } as PermissionRule<TResource, TQuery, TResourceType, TParams>;
   }
   return rule;
 }

